File: c#/src/v1/CommsSdk.cs
using Comms.Models;
using Comms.Utils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace Comms
{
    public class CommsSdk
    {
        public static string ApiUrl { get; private set; } = "https://comms.egosms.co/api/v1/json/";
        private static readonly HttpClient Client = new();

        public string? UserName { get; private set; }
        public string? ApiKey { get; private set; }
        public string SenderId { get; private set; } = "EgoSMS";
        public bool IsAuthenticated { get; set; }

        private CommsSdk() { }

        public static async Task<CommsSdk> Authenticate(string userName, string apiKey)
        {
            var sdk = new CommsSdk
            {
                UserName = userName,
                ApiKey = apiKey
            };
            await Validator.ValidateCredentials(sdk);
            return sdk;
        }

        public static void UseSandBox()
        {
            ApiUrl = "https://comms-test.pahappa.net/api/v1/json";
        }

        public static void UseLiveServer()
        {
            ApiUrl = "https://comms.egosms.co/api/v1/json";
        }

        public void SetAuthenticated()
        {
            IsAuthenticated = true;
        }

        public CommsSdk WithSenderId(string senderId)
        {
            SenderId = senderId;
            return this;
        }

        public Task<bool> SendSms(string number, string message) =>
            SendSms(new List<string> { number }, message, SenderId, MessagePriority.Highest);

        public Task<bool> SendSms(string number, string message, string senderId) =>
            SendSms(new List<string> { number }, message, senderId, MessagePriority.Highest);

        public Task<bool> SendSms(string number, string message, string senderId, MessagePriority priority) =>
            SendSms(new List<string> { number }, message, senderId, priority);

        public Task<bool> SendSms(string number, string message, MessagePriority priority) =>
            SendSms(new List<string> { number }, message, SenderId, priority);

        public Task<bool> SendSms(List<string> numbers, string message) =>
            SendSms(numbers, message, SenderId, MessagePriority.Highest);

        public Task<bool> SendSms(List<string> numbers, string message, string senderId) =>
            SendSms(numbers, message, senderId, MessagePriority.Highest);

        public Task<bool> SendSms(List<string> numbers, string message, MessagePriority priority) =>
            SendSms(numbers, message, SenderId, priority);

        public async Task<bool> SendSms(List<string> numbers, string message, string senderId, MessagePriority priority)
        {
            var apiResponse = await QuerySendSms(numbers, message, senderId, priority);
            if (apiResponse == null)
            {
                Console.WriteLine("Failed to get a response from the server.");
                return false;
            }

            switch (apiResponse.Status)
            {
                case ApiResponseCode.OK:
                    Console.WriteLine("SMS sent successfully.");
                    Console.WriteLine($"MessageFollowUpUniqueCode: {apiResponse.MessageFollowUpCode}");
                    return true;
                case ApiResponseCode.Failed:
                    Console.WriteLine($"Failed: {apiResponse.Message}");
                    return false;
                default:
                    throw new Exception("Unexpected response status: " + apiResponse.Status);
            }
        }

        /// <summary>Same as <see cref="SendSms"/> but returns the full <see cref="ApiResponse"/> object.</summary>
        public async Task<ApiResponse?> QuerySendSms(List<string> numbers, string message, string senderId, MessagePriority priority)
        {
            if (await SdkNotAuthenticated()) return null;

            if (numbers == null || !numbers.Any())
                throw new ArgumentException("Numbers list cannot be empty");

            if (string.IsNullOrEmpty(message))
                throw new ArgumentException("Message cannot be empty");

            if (message.Length == 1)
                throw new ArgumentException("Message cannot be a single character");

            if (string.IsNullOrWhiteSpace(senderId))
                senderId = SenderId;

            if (senderId?.Length > 11)
                Console.WriteLine("Warning: Sender ID length exceeds 11 characters. Some networks may truncate or reject messages.");

            numbers = NumberValidator.ValidateNumbers(numbers);
            if (!numbers.Any())
            {
                Console.Error.WriteLine("No valid phone numbers provided. Please check inputs.");
                return null;
            }

            var apiRequest = new ApiRequest
            {
                Method = "SendSms",
                MessageData = numbers.Select(num => new MessageModel(num, message, senderId ?? this.SenderId, priority)).ToList(),
                Userdata = new UserData(UserName!, ApiKey!),
            };

            try
            {
                var json = JsonSerializer.Serialize(apiRequest);
                var content = new StringContent(json, Encoding.UTF8, "application/json");
                var response = await Client.PostAsync(ApiUrl, content);
                response.EnsureSuccessStatusCode();

                var responseBody = await response.Content.ReadAsStringAsync();
                return JsonSerializer.Deserialize<ApiResponse>(responseBody);
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Failed to send SMS: " + e.Message);
                try
                {
                    Console.Error.WriteLine("Request: " + JsonSerializer.Serialize(apiRequest));
                }
                catch { }
                return null;
            }
        }

        private async Task<bool> SdkNotAuthenticated()
        {
            if (IsAuthenticated) return false;

            Console.Error.WriteLine("SDK is not authenticated. Please authenticate before performing actions.");
            Console.Error.WriteLine("Attempting to re-authenticate with provided credentials...");
            return !await Validator.ValidateCredentials(this);
        }

        /// <summary>Same as <see cref="GetBalance"/> but returns the full <see cref="ApiResponse"/> object.</summary>
        public async Task<ApiResponse?> QueryBalance()
        {
            if (await SdkNotAuthenticated()) return null;

            var apiRequest = new ApiRequest
            {
                Method = "Balance",
                Userdata = new UserData(UserName!, ApiKey!)
            };

            try
            {
                var json = JsonSerializer.Serialize(apiRequest);
                var content = new StringContent(json, Encoding.UTF8, "application/json");
                var response = await Client.PostAsync(ApiUrl, content);
                response.EnsureSuccessStatusCode();

                var responseBody = await response.Content.ReadAsStringAsync();
                return JsonSerializer.Deserialize<ApiResponse>(responseBody);
            }
            catch (Exception e)
            {
                throw new Exception("Failed to get balance: " + e.Message, e);
            }
        }

        public async Task<double?> GetBalance()
        {
            var response = await QueryBalance();
            return response?.Balance;
        }

        public override string ToString()
        {
            return $"SDK({UserName} => {ApiKey})";
        }
    }

}

File: dart/lib/src/v1/comms_sdk.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'models/api_request.dart';
import 'models/api_response.dart';
import 'models/message_model.dart';
import 'models/message_priority.dart';
import 'models/user_data.dart';
import 'utils/number_validator.dart';
import 'utils/validator.dart';

class CommsSDK {
  static String _apiUrl = 'https://comms.egosms.co/api/v1/json/';
  static String getApiUrl() => _apiUrl;

  String? userName;
  String? apiKey;

  String senderId = 'EgoSMS';
  bool isAuthenticated = false;
  final http.Client _client = http.Client();

  CommsSDK._();

  static Future<CommsSDK> authenticate(
    String userName,
    String apiKey,
  ) async {
    final sdk = CommsSDK._();
    sdk.userName = userName;
    sdk.apiKey = apiKey;
    await Validator.validateCredentials(sdk);
    return sdk;
  }

  static void useSandBox() {
    _apiUrl = 'https://comms-test.pahappa.net/api/v1/json';
  }

  static void useLiveServer() {
    _apiUrl = 'https://comms.egosms.co/api/v1/json';
  }

  void setAuthenticated() {
    isAuthenticated = true;
  }

  CommsSDK withSenderId(String senderId) {
    this.senderId = senderId;
    return this;
  }

  Future<bool> sendSMS({
    required List<String> numbers,
    required String message,
    String? senderId,
    MessagePriority? priority,
  }) async {
    final apiResponse = await querySendSMS(
      numbers: numbers,
      message: message,
      senderId: senderId ?? this.senderId,
      priority: priority ?? MessagePriority.HIGHEST,
    );

    if (apiResponse == null) {
      print('Failed to get a response from the server.');
      return false;
    }

    if (apiResponse.status.name.toLowerCase() == "ok") {
      print('SMS sent successfully.');
      print('MessageFollowUpUniqueCode: ${apiResponse.messageFollowUpCode}');
      return true;
    } else if (apiResponse.status.name.toLowerCase() == "failed") {
      print('Failed: ${apiResponse.message}');
      return false;
    } else {
      throw Exception('Unexpected response status: ${apiResponse.status}');
    }
  }

  /// Same as [sendSMS] but returns the full [ApiResponse] object.
  Future<ApiResponse?> querySendSMS({
    required List<String> numbers,
    required String message,
    required String senderId,
    required MessagePriority priority,
  }) async {
    if (await _sdkNotAuthenticated()) return null;
    if (numbers.isEmpty) {
      throw ArgumentError('Numbers list cannot be empty');
    }
    if (message.isEmpty) {
      throw ArgumentError('Message cannot be empty');
    }
    if (message.length == 1) {
      throw ArgumentError('Message cannot be a single character');
    }
    if (senderId.trim().isEmpty) {
      senderId = this.senderId;
    }
    if (senderId.length > 11) {
      print('Warning: Sender ID length exceeds 11 characters. Some networks may truncate or reject messages.');
    }

    numbers = NumberValidator.validateNumbers(numbers);
    if (numbers.isEmpty) {
      print('No valid phone numbers provided. Please check inputs.');
      return null;
    }

    final messageModels = numbers
        .map(
          (number) => MessageModel(
            number: number,
            message: message,
            senderId: senderId,
            priority: priority,
          ),
        )
        .toList();

    final apiRequest = ApiRequest(
      method: 'SendSms',
      messageData: messageModels,
      userdata: UserData(userName!, apiKey!),
    );

    try {
      final res = await _client.post(
        Uri.parse(_apiUrl),
        body: jsonEncode(apiRequest.toJson()),
        headers: {'Content-Type': 'application/json'},
      );
      return ApiResponse.fromJson(jsonDecode(res.body));
    } catch (e) {
      print('Failed to send SMS: $e');
      print('Request: ${jsonEncode(apiRequest.toJson())}');
      return null;
    }
  }

  Future<bool> _sdkNotAuthenticated() async {
    if (!isAuthenticated) {
      print(
        'SDK is not authenticated. Please authenticate before performing actions.',
      );
      print('Attempting to re-authenticate with provided credentials...');
      return !await Validator.validateCredentials(this);
    }
    return false;
  }

  /// Same as [getBalance] but returns the full [ApiResponse] object.
  Future<ApiResponse?> queryBalance() async {
    if (await _sdkNotAuthenticated()) {
      return null;
    }
    final apiRequest = ApiRequest(
      method: 'Balance',
      userdata: UserData(userName!, apiKey!),
      messageData: [],
    );
    try {
      final res = await _client.post(
        Uri.parse(_apiUrl),
        body: jsonEncode(apiRequest.toJson()),
        headers: {'Content-Type': 'application/json'},
      );
      return ApiResponse.fromJson(jsonDecode(res.body));
    } catch (e) {
      throw Exception('Failed to get balance: $e');
    }
  }

  Future<double?> getBalance() async {
    final response = await queryBalance();
    return response?.balance != null ? double.tryParse(response!.balance!) : null;
  }

  @override
  String toString() {
    return 'SDK($userName => $apiKey)';
  }
}
